File: ParallelHuffmanCoding/cpp/ParallelHuffmanCoding.cpp
#include <fstream>
#include <sstream>
#include <string>
#include <cerrno>
#include <iostream>
#include <queue>          // priority_queue
#include <vector>         // vector
#include <functional>
#include <chrono>
#include <future>
#include <pthread.h>

using namespace std;

class Node {
    public:
        char ch;
        int freq;
        Node* left; 
        Node* right;

        Node(char ch, int freq, Node* left, Node* right) {
            this->ch = ch;
            this->freq = freq;
            this->left = left;
            this->right = right;
        }

        // is the node a leaf node?
        bool isLeaf() {
            return (left == 0) && (right == 0);
        }
};

class CompressedOutput {
    public:
        vector<bool>* compressed;
        int inputLength;
};

string get_file_contents(const char *filename)
{
    ifstream in(filename, ios::in | ios::binary);
    if (in)
    {
        ostringstream contents;
        contents << in.rdbuf();
        in.close();
        return (contents.str());
    }
    throw(errno);
}

bool compare (Node* lhs, Node* rhs)
{
    return (lhs->freq > rhs->freq);
}

Node* buildTrie(int* freq) {
    priority_queue<Node*,vector<Node*>, function<bool(Node*,Node*)>> pq(compare);

    for (int c = 0; c < 256; c++) {
        if (freq[c] > 0) {
            pq.push(new Node(c, freq[c], 0, 0));
        }
    }

    while(pq.size() > 1) {
        Node* left = pq.top();
        pq.pop();
        Node* right = pq.top();
        pq.pop();
        pq.push(new Node('0', left->freq + right->freq, left, right));
    }

    return pq.top();
}

void buildCode(string* st, Node* x, string s) {
    if (!(x->isLeaf())) {
        buildCode(st, x->left, s + '0');
        buildCode(st, x->right, s + '1');
    } else {
        st[x->ch] = s;
    }
}

CompressedOutput compress(const char * chars, int index, int inputLength, string* st) {
    vector<bool>* compressed = new vector<bool>();

    float div8 = float(inputLength) / 8;
    int start = index * int(div8);
    int end = (index + 1) * int(div8);
    if (index == 7) {
        end = inputLength;
    }
    for (int i = start; i < end; i++) {
        string code = st[chars[i]];
        const char * codeCh = code.c_str();
        for (int j = 0; j < code.length(); j++) {
            if (codeCh[j] == '0') {
                (compressed)->push_back(false);
            } else if (codeCh[j] == '1') {
                (compressed)->push_back(true);
            }
        }
    }
    cout << "Input size:" << end - start << " Compressed:" << (compressed)->size()/8 << endl;

    CompressedOutput retVal;
    retVal.compressed = compressed;
    retVal.inputLength = end - start;
    return retVal;
}

void buildTrieAndCode(const char * chars, int inputLength, Node** root, string** st) {
    // tabulate frequency counts
    int freq[256] {0};
    for (int i = 0; i < inputLength; i++) {
        freq[chars[i]]++;
    }

    *root = buildTrie(freq);

    *st = new string[256];
    buildCode(*st, *root, "");
}

string expand(int inputLength, Node* root, vector<bool>* compressed) {
    char* output = new char[inputLength + 1];

    int j = 0;
    for (int i = 0; i < inputLength; i++) {
        Node * x = root;
        while(!(x->isLeaf())) {
            bool bit = compressed->at(j++);
            if (bit) {
                x = x->right;
            } else {
                x = x->left;
            }
        }
        output[i] = x->ch;
    }
    output[inputLength] = '\0';
    return string(output);
}

void cleanupNode(Node * x) {
    if (x->left != 0) {
        cleanupNode(x->left);
        x->left = 0;
    }
    if (x->right != 0) {
        cleanupNode(x->right);
        x->right = 0;
    }
    delete x;
}

void printTime(chrono::steady_clock::time_point begin, string text) {
    chrono::steady_clock::time_point endFile = chrono::steady_clock::now();
	cout << text << chrono::duration_cast<chrono::microseconds>(endFile - begin).count() << "[µs]" << endl;
}

vector<CompressedOutput> compressParallel(const char * chars, int inputLength, Node** root, string** st) {
    buildTrieAndCode(chars, inputLength, root,st);
    vector<future<CompressedOutput>> futures;
    for (int i = 0; i < 8; i++) {
        futures.push_back(async(compress, chars, i, inputLength, *st));
    }
    vector<CompressedOutput> compressedOutputs;
    for (int i = 0; i < 8; i++) {
        compressedOutputs.push_back(futures.at(i).get());
    }
    return compressedOutputs;
}

void expandParallel(vector<CompressedOutput> compressedOutputs, Node* root, string fileContent) {
    vector<future<string>> expandFutures;
    for (int i = 0; i < 8; i++) {
        expandFutures.push_back(async(expand, compressedOutputs.at(i).inputLength, root, compressedOutputs.at(i).compressed));
    }
    string expanded;
    for (int i = 0; i < 8; i++) {
        expanded += expandFutures.at(i).get();
    }
    cout << fileContent.length() << ' ' << expanded.length() << endl;
    if (fileContent.compare(expanded) == 0) {
        cout << "Success" << endl;
    }
}

int main()
{
    Node* root;
    vector<bool>* compressed;
    string* st;
    chrono::steady_clock::time_point begin = chrono::steady_clock::now();

    string fileContent = get_file_contents("bible.txt");
    const char * chars = fileContent.c_str();
    int inputLength = fileContent.length();
    printTime(begin, "Time File Read = ");
    vector<CompressedOutput> compressedOutputs = compressParallel(chars, inputLength, &root, &st);
    printTime(begin, "Time Compress = ");
    expandParallel(compressedOutputs, root, fileContent);
    printTime(begin, "Time difference = ");

    delete compressed;
    cleanupNode(root);
}

File: ParallelHuffmanCoding/java/ParallelHuffmanCoding.java
import java.util.*;
import java.util.stream.*;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import java.io.*;
import java.time.*;

public class ParallelHuffmanCoding {
    // alphabet size of extended ASCII
    private static final int R = 256;

    private Node trie;

    private int length;

    private String fileContent;

    // Huffman trie node
    private static class Node implements Comparable<Node> {
        private final char ch;
        private final int freq;
        private final Node left, right;

        Node(char ch, int freq, Node left, Node right) {
            this.ch = ch;
            this.freq = freq;
            this.left = left;
            this.right = right;
        }

        // is the node a leaf node?
        private boolean isLeaf() {
            assert ((left == null) && (right == null)) || ((left != null) && (right != null));
            return (left == null) && (right == null);
        }

        // compare, based on frequency
        public int compareTo(Node that) {
            return this.freq - that.freq;
        }
    }

    private static class CompressedOutput {
        ArrayList<Boolean> compressed;
        int inputLength;

        CompressedOutput(ArrayList<Boolean> c, int i) {
            compressed = c;
            inputLength = i;
        }
    }

    private String readFromInputStream(InputStream inputStream) throws IOException {
        StringBuilder resultStringBuilder = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream))) {
            String line;
            while ((line = br.readLine()) != null) {
                resultStringBuilder.append(line).append("\n");
            }
        }
        return resultStringBuilder.toString();
    }

    private String readFile(String name) {
        try {
            File file = new File(name);
            if (file.exists()) {
                FileInputStream fis = new FileInputStream(file);
                return readFromInputStream(fis);
            }
        } catch (IOException ioe) {
            System.err.println("Could not open " + name);
        }

        return "";
    }

    /**
     * Reads a sequence of 8-bit bytes from standard input; compresses them using
     * Huffman codes with an 8-bit alphabet; and writes the results to standard
     * output.
     */
    public CompressedOutput compress(char[] input, String[] st, int index) {
        
        ArrayList<Boolean> compressed = new ArrayList<>();

        float div8 = (float)length / 8;
        int start = index * (int)div8;
        int end = (index + 1) * (int)div8;
        if (index == 7) {
            end = length;
        }

        for (int i = start; i < end; i++) {
            String code = st[input[i]];
            for (int j = 0; j < code.length(); j++) {
                if (code.charAt(j) == '0') {
                    compressed.add(false);
                } else if (code.charAt(j) == '1') {
                    compressed.add(true);
                } else
                    throw new IllegalStateException("Illegal state");
            }
        }

        System.out.println("Input length:" + (end - start) + " Compressed:" + compressed.size() / 8);
        return new CompressedOutput(compressed, end - start);
    }

    public List<CompressedOutput> compressParallel(String name) {
        // read the input
        String s = readFile(name);
        fileContent = s;
        char[] input = s.toCharArray();

        // tabulate frequency counts
        int[] freq = new int[R];
        for (int i = 0; i < input.length; i++)
            freq[input[i]]++;

        // build Huffman trie
        Node root = buildTrie(freq);
        trie = root;

        // build code table
        String[] st = new String[R];
        buildCode(st, root, "");

        length = input.length;

        List<CompletableFuture<CompressedOutput>> compressFutures = new ArrayList<>();
        
        IntStream.range(0, 8).forEach(i -> {
            compressFutures.add(CompletableFuture.supplyAsync(() -> compress(input, st, i)));
        });

        return compressFutures.stream().map(f -> getFuture(f)).collect(Collectors.toList());
    }

    static <T> T getFuture(CompletableFuture<T> f) {
        try {
            return f.get();
        } catch (Exception ex) {

        }
        return null;
    }

    // build the Huffman trie given frequencies
    private Node buildTrie(int[] freq) {

        // initialze priority queue with singleton trees
        PriorityQueue<Node> pq = new PriorityQueue<>();
        for (char c = 0; c < R; c++) {
            if (freq[c] > 0) {
                pq.add(new Node(c, freq[c], null, null));
            }
        }

        // merge two smallest trees
        while (pq.size() > 1) {
            Node left = pq.remove();
            Node right = pq.remove();
            Node parent = new Node('\0', left.freq + right.freq, left, right);
            pq.add(parent);
        }
        return pq.remove();
    }

    // make a lookup table from symbols and their encodings
    private void buildCode(String[] st, Node x, String s) {
        if (!x.isLeaf()) {
            buildCode(st, x.left, s + '0');
            buildCode(st, x.right, s + '1');
        } else {
            st[x.ch] = s;
        }
    }

    /**
     * Reads a sequence of bits that represents a Huffman-compressed message from
     * standard input; expands them; and writes the results to standard output.
     */
    public String expand(CompressedOutput input) {

        // read in Huffman trie from input stream
        Node root = trie;

        Iterator<Boolean> iter = input.compressed.iterator();
        StringBuilder output = new StringBuilder();

        // decode using the Huffman trie
        for (int i = 0; i < input.inputLength; i++) {
            Node x = root;
            while (!x.isLeaf()) {
                boolean bit = iter.next();// BinaryStdIn.readBoolean();
                if (bit)
                    x = x.right;
                else
                    x = x.left;
            }
            output.append(x.ch);
        }
        return output.toString();
    }

    public String expandParallel(List<CompressedOutput> inputs) {
        List<CompletableFuture<String>> futures = new ArrayList<>();

        futures = inputs.stream().map(i -> CompletableFuture.supplyAsync(() -> expand(i))).collect(Collectors.toList());

        return futures.stream().map(f -> getFuture(f)).collect(StringBuilder::new, StringBuilder::append, StringBuilder::append).toString();
    }

    /**
     * Sample client that calls {@code compress()} if the command-line argument is
     * "-" an {@code expand()} if it is "+".
     *
     * @param args the command-line arguments
     */
    public static void main(String[] args) {
        ParallelHuffmanCoding hc = new ParallelHuffmanCoding();
        Instant start = Instant.now();
        List<CompressedOutput> inputs = hc.compressParallel("bible.txt");
        String expanded = hc.expandParallel(inputs);
        if (!hc.fileContent.equals(expanded))
            throw new IllegalArgumentException("Strings dont match");
        Instant end = Instant.now();
        System.out.println(Duration.between(start, end));
    }
}
